### ISSUE TITLE
Inconsistency between NT-ASGD described in the research paper and main.py

### ISSUE THREAD

**Issue by takase** [salesforce/awd-lstm-lm#13]
The NT-ASGD algorithm compares the validation loss value with previous _n_ loss values, but I think main.py compares the validation loss value with loss values which are from 1st epoch to _t-n_ epoch because of line 222:
`if 't0' not in optimizer.param_groups[0] and (len(best_val_loss)>args.nonmono and val_loss > min(best_val_loss[:-args.nonmono])):`
If the line is revised to
`if 't0' not in optimizer.param_groups[0] and (len(best_val_loss)>args.nonmono and val_loss > min(best_val_loss[-args.nonmono:])):`
the line is consistent with the NT-ASGD described in the research paper.

However, if we use the line, the code starts averaging immediately after the validation metric worsens.
So, what about using the following line?
`if 't0' not in optimizer.param_groups[0] and (len(best_val_loss)>args.nonmono and val_loss > max(best_val_loss[-args.nonmono:])):`

**Comment by keskarnitish** [comment]
We apologize for the confusion; the paper has a typo w.r.t. the triggering condition. The condition in the code is correct; we will update the paper soon. The condition is designed such that it ignores/masks the last `args.nonmono` epochs and compares the best value of the metric in this set with the current value. There are other ways to design a non-monotone trigger but we found that this ensures that the learning rate is reduced in a timely fashion without making it too frequent. Hope this helps!
